src/Adapter/MSTest.TestAdapter/Discovery/AssemblyEnumerator.cs

```diff
@@ -409,25 +409,26 @@ private static bool ProcessTestDataSourceTests(UnitTestElement test, ReflectionT
 
             // This code is to discover tests. To run the tests code is in TestMethodRunner.ExecuteDataSourceBasedTests.
             // Any change made here should be reflected in TestMethodRunner.ExecuteDataSourceBasedTests as well.
-            try
+            data = dataSource.GetData(methodInfo);
+
+            if (!data.Any())
             {
-                data = dataSource.GetData(methodInfo);
+                if (MSTestSettings.CurrentSettings.ConsiderEmptyDataSourceAsInconclusive)
```

Comment:
Optional nit: I usually to prefer exit conditions first to reduce mental overhead so I would suggest to revert the if to throw and not to have some else.